hands-on-exercise/4-run-in-prod/3-migration-prod.md

```diff
@@ -0,0 +1,616 @@
+    +  FROM --platform=linux checkersd_i:v2-alpine AS v2
+        FROM --platform=linux alpine
+        ...
+        COPY --from=builder /root/cosmos-sdk/cosmovisor/cosmovisor ${LOCAL}/bin/cosmovisor
+    +  COPY --from=v1 /usr/local/bin/checkersd $DAEMON_HOME/cosmovisor/genesis/bin/checkersd
+    +  COPY --from=v2 /usr/local/bin/checkersd $DAEMON_HOME/cosmovisor/upgrades/v1tov2/bin/checkersd
+        ...
+    ```
+
+    It starts at v1, therefore the v1 code goes into `.../genesis`, and we know that the code of the eventual upgrade named `v1tov2` is the v2. Note also the decision to use `/usr/local` explicitly as this is knowledge that is kept in a separate Docker image.
+
+4. And make Cosmovisor start by default:
+
+    ```diff [https://github.com/cosmos/b9-checkers-academy-draft/blob/migration-prod/prod-sim/Dockerfile-cosmovisor-alpine#L29]
+        COPY --from=v2 /usr/local/bin/checkersd $DAEMON_HOME/cosmovisor/upgrades/v1tov2/bin/checkersd
+
+    +  ENTRYPOINT [ "cosmovisor" ]
+    ```
+
+When you put it all together, you get:
+
+```Dockerfile [https://github.com/cosmos/b9-checkers-academy-draft/blob/migration-prod/prod-sim/Dockerfile-cosmovisor-alpine]
+FROM --platform=linux golang:1.18.7-alpine AS builder
+
+ENV COSMOS_VERSION=v0.45.4
+
+RUN apk update
+RUN apk add make git
+
+WORKDIR /root
+RUN git clone --depth 1 --branch ${COSMOS_VERSION} https://github.com/cosmos/cosmos-sdk.git
+
+WORKDIR /root/cosmos-sdk/cosmovisor
+
+RUN make cosmovisor
+
+FROM --platform=linux checkersd_i:v1-alpine AS v1
+FROM --platform=linux checkersd_i:v2-alpine AS v2
+FROM --platform=linux alpine
+
+ENV LOCAL=/usr/local
+ENV DAEMON_HOME=/root/.checkers-upgrade
+ENV DAEMON_NAME=checkersd
+ENV DAEMON_ALLOW_DOWNLOAD_BINARIES=false
+ENV DAEMON_RESTART_AFTER_UPGRADE=true
+
+COPY --from=builder /root/cosmos-sdk/cosmovisor/cosmovisor ${LOCAL}/bin/cosmovisor
+COPY --from=v1 /usr/local/bin/checkersd $DAEMON_HOME/cosmovisor/genesis/bin/checkersd
+COPY --from=v2 /usr/local/bin/checkersd $DAEMON_HOME/cosmovisor/upgrades/v1tov2/bin/checkersd
+
+ENTRYPOINT [ "cosmovisor" ]
+```
+
+Now you can create the Cosmovisor Docker image with a meaningful tag:
+
+```sh
+$ docker build . \
+    -f prod-sim/Dockerfile-cosmovisor-alpine \
+    -t cosmovisor_i:v1tov2-alpine
+```
+
+## Docker Compose elements
+
+With the executables and the blockchain elements ready, you can now define the _production setup_. You already defined one in the previous [run checkers in prod section](/hands-on-exercise/4-run-in-prod/1-run-prod-docker.md). In this new setup, the only things that change are the Docker images you call: `cosmovisor_i` instead of `checkersd_i`. Even the `start` command does not need to change.
+
+To avoid rewriting everything, you can declare a Docker Compose [extension](https://docs.docker.com/compose/extends/) in a new file `prod-sim/docker-compose-cosmovisor.yml`. Each `checkersd` type of service is extended, and in fine replaced, with a new `image`:
```

Comment:
@xavierlepretre 
> Each `checkersd` type of service is extended, and **_in fine_** replaced, with a new `image`:
Is "in fine" a typo of "in **time**"?

Reply:
No as in "finally", "in the end".